js/model.js

```diff
@@ -7,21 +7,165 @@ const LETTER_FREQUENCY = [
+ * endPoint
+ * @param {!Point} startPoint
+ * @param {!Point} endPoint
+ * @param {!Point} targetPoint
+ * @returns {!Point}
+ */
+function getNearestPointOnLine(startPoint, endPoint, targetPoint) {
+  let v = endPoint.subtract(startPoint);
+
+  let u = targetPoint.subtract(startPoint);
+
+  let s = u.dot(v) / v.dot(v);
+
+  if (s < 0) {
+    return startPoint;
+  } else if (s > 1) {
+    return endPoint;
+  } else {
+    let I = startPoint;
+    let w = new Point(s * v.x, s * v.y);
+    const result = new Point(I.x + w.x, I.y + w.y);
+    return result;
+  }
+}
+
 class GameBoard {
-  constructor() {
+  /**
+   * @param {number} startPointX
+   * @param {number} startPointY
+   * @param {number} gameHeight
+   * @param {number} gameWidth
+   * @param {number} dimensionsOfSquare
+   * @param {!Array<!Array<!Square>>} board
+   */
+  constructor(
+      startPointX, startPointY, gameHeight, gameWidth, dimensionsOfSquare,
+      board) {
     /** @type {number} */
-    this.startPointX;
+    this.startPointX = startPointX;
     /** @type {number} */
-    this.startPointY;
+    this.startPointY = startPointY;
     /** @type {number} */
-    this.gameHeight;
+    this.gameHeight = gameHeight;
     /** @type {number} */
-    this.gameWidth;
+    this.gameWidth = gameWidth;
     /** @type {number} */
-    this.dimensionsOfSquare;
+    this.dimensionsOfSquare = dimensionsOfSquare;
     /** @type {!Array<!Array<!Square>>} */
-    this.board;
+    this.board = board;
+  }
+
+  /**
+   * Returns the square that contains the coords x, y
+   * @param {!Point} point
+   * @returns {?Square}
+   */
+  findSquare(point) {
+    for (const row of this.board) {
+      for (const square of row) {
+        const widthOfSquare = square.xCoord + this.dimensionsOfSquare;
+        const heightOfSquare = square.yCoord + this.dimensionsOfSquare;
+        if (point.x <= widthOfSquare && point.y <= heightOfSquare &&
+            point.x > square.xCoord && point.y > square.yCoord) {
+          return square;
+        }
+      }
+    }
+    return null;  // TODO - Handle case where the mouse is off the screen
+  }
+
+
+  /**
+   * Determines if the coords x, y lie within the game board
+   * @param {!Point} point
+   * @returns {boolean}
+   */
+  isOnBoard(point) {
+    const widthOfBoard =
+        this.startPointX + this.gameWidth * this.dimensionsOfSquare;
+    const heightOfBoard =
+        this.startPointY + this.gameHeight * this.dimensionsOfSquare;
+
+    if (point.x < widthOfBoard && point.y < heightOfBoard &&
+        point.x > this.startPointX && point.y > this.startPointY) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Gets all the squares that lie within a distance of 2 pixels from the line
+   * created by the parameters to the center of each square
+   * @param {!Point} startPoint
+   * @param {!Point} endPoint
+   * @returns {!Array<!Square>}
+   */
+  getSquaresFromLine(startPoint, endPoint) {
```

Comment:
I don't have a good idea, but feel like there must be more optimal algorithm than iterating through all squares. Don't need to address it now but let's add a TODO for this.

Reply:
I saw this comment and thought the same thing. I was interested in improving on it so I came up with a different solution that makes use of right angle triangles. It actually makes the code to for `getNearestPointOnLine` irrelevant. The solution also doesn't iterate through all the points, the only downfall of the more optimal algorithm is that it works more with floating point numbers which required me to do a lot of 'epsilon' checks for accuracy. I left both in the code with the old one commented out. (It also solves another issue I was having with how to account for a search area that shouldn't be allowed.